src/main/java/org/candlepin/resource/util/KeyValueStringParser.java

```diff
@@ -0,0 +1,271 @@
+/**
+ * Copyright (c) 2009 - 2021 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.resource.util;
+
+import org.candlepin.exceptions.BadRequestException;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+
+
+
+/**
+ * Provides utility function for parsing key-value strings received from clients.
+ * <p></p>
+ * The functions provided by this class are intended to be used by resource-level or user-facing
+ * business functions. As such, they are very lenient on input, accepting null or empty collections,
+ * and only raising exceptions when data is both present and malformed. In the case of an exception,
+ * the exceptions are intended to be passed through to the user and contain both the appropriate
+ * HTTP error code and translated error message. Use of these functions should not require
+ * additional error handling or input validation in the general case.
+ * <p></p>
+ * Key-value strings are parsed according to the format "{key}:{value}", where only the first colon
+ * is recognized as a delimiter, and all additional colons will be rolled into the value. If the
+ * string is null, empty, or otherwise does not contain any instances of the delimiter, it is
+ * considered malformed, and will cause functions in this class to throw a BadRequestException.
+ */
+public class KeyValueStringParser {
```

Comment:
Could we turn this into normal utility class with constructor injection? Static injections are greatly discouraged even by the guice team. They are inconvenient to test and will make maintenance of our API layer difficult.

Reply:
I've got a commit prepped where this change has been made, but I don't see what we gain from it outside of avoiding static injection. The KeyValueStringParser is stateless by design, so the change ends up going from:

```java
KeyValueStringParser.parseKeyValuePairs(attrFilters)
  .forEach(kvpair -> poolFilters.addAttributeFilter(kvpair.getKey(), kvpair.getValue()));
```

to:

```java
new KeyValueStringParser(this.i18n).parseKeyValuePairs(attrFilters)
  .forEach(kvpair -> poolFilters.addAttributeFilter(kvpair.getKey(), kvpair.getValue()));
```

which is basically the same thing, but with manual construction of an object that is immediately discarded.

We could inject a provider to each resource that uses it, which would change it to:

```java

@Inject
public OwnerResource(..., Provider<KeyValueStringParser> kvsParserProvider, ...) {
  ...

  this.kvsParserProvider = Objects.requireNonNull(kvsParserProvider);
  // Alternatively, this could just be the KeyValueStringProvider directly
}

...

this.kvStringParserProvider[.get()].parseKeyValuePairs(attrFilters)
  .forEach(kvpair -> poolFilters.addAttributeFilter(kvpair.getKey(), kvpair.getValue()));
```

This works, but now we need to update the constructor so we accept and store a reference to a utility that, again, isn't storing any state and is, effectively an implementation detail. We're using a composition pattern to inject fixed behavior to avoid injection (which really is already just avoiding making the resources have to handle and translate an exception message).

And then the tests need to be updated to pass in the provider/instance on invocation. But the core idea is to *not* have different implementations of that parsing, so I don't see why we would want control over those details or need explicit testing around knowledge of the calls into this utility outside of explicit testing of it in the targeted suite.

Something I had considered to ease any static injection pain is having a default i18n instance as the default, and let static injection override it:

```java
    private static Provider<I18n> i18nProvider = () -> 
        I18nFactory.getI18n(KeyValueStringParser.class, Locale.US, I18nFactory.FALLBACK);

    @Inject
    public static Provider<I18n> initialize(Provider<I18n> i18nProvider) {
        if (i18nProvider == null) {
            throw new IllegalArgumentException("i18nProvider is null");
        }

        Provider<I18n> prev = KeyValueStringParser.i18nProvider;
        KeyValueStringParser.i18nProvider = i18nProvider;

        return prev;
    }
```

This would get rid of any test injector setup (though it's only needed in the test suite for this class), and allow for one-off changes if necessary.

Is there something I'm overlooking here? Any other thoughts on it?

Reply:
Creating the instance where needed could work. It's just a small utility class so we can justify it not being injected into resources. The main issue is that static injection creates a dependency on Guice which will get in our way when we'll want to migrate to different injection framework.

Reply:
ack; will convert to non-static utility and locally created instances

Reply:
done